src/invariant/test/ConservationOfLumensTests.cpp

```diff
@@ -301,3 +306,260 @@ TEST_CASE("Inflation changes are consistent",
+            ltx.commit();
+        }
+
+        ::closeLedger(*app);
+
+        app->getInvariantManager().enableInvariant("ConservationOfLumens");
+
+        auto ledgerState =
+            app->getLedgerManager().getLastClosedLedgerStateForTesting();
+        auto& inMemoryState =
+            app->getLedgerManager().getInMemorySorobanStateForTesting();
+
+        REQUIRE_NOTHROW(app->getInvariantManager().runStateSnapshotInvariant(
+            ledgerState, inMemoryState, []() { return false; }));
+    }
+
+    SECTION("Invariant fails when bucket balance doesn't match totalCoins")
+    {
+        // Get initial totalCoins before adding test accounts
+        int64_t initialTotal = app->getLedgerManager()
+                                   .getLastClosedLedgerHeader()
+                                   .header.totalCoins;
+
+        // Add entries to live bucket
+        lm.setNextLedgerEntryBatchForBucketTesting({acc1, acc2}, // init entries
+                                                   {},           // live entries
+                                                   {});          // dead entries
+        ::closeLedger(*app);
+
+        // Set totalCoins to incorrect value (off by 9999)
+        int64_t incorrectTotal = initialTotal + acc1.data.account().balance +
+                                 acc2.data.account().balance + 9999;
+
+        {
+            LedgerTxn ltx(app->getLedgerTxnRoot());
+            ltx.loadHeader().current().totalCoins = incorrectTotal;
+            ltx.commit();
+        }
+
+        ::closeLedger(*app);
+
+        app->getInvariantManager().enableInvariant("ConservationOfLumens");
+
+        auto ledgerState =
+            app->getLedgerManager().getLastClosedLedgerStateForTesting();
+        auto& inMemoryState =
+            app->getLedgerManager().getInMemorySorobanStateForTesting();
+
+        REQUIRE_THROWS_AS(
+            app->getInvariantManager().runStateSnapshotInvariant(
+                ledgerState, inMemoryState, []() { return false; }),
+            InvariantDoesNotHold);
+    }
+
+    SECTION("Invariant handles shadowing correctly")
+    {
+        // Get initial totalCoins
+        int64_t initialTotal = app->getLedgerManager()
+                                   .getLastClosedLedgerHeader()
+                                   .header.totalCoins;
+
+        // Add acc1 as INIT entry in older level
+        lm.setNextLedgerEntryBatchForBucketTesting({acc1}, // init entries
+                                                   {},     // live entries
+                                                   {});    // dead entries
+        ::closeLedger(*app);
+
+        // Create modified version of acc1 with different balance
+        auto acc1Modified = acc1;
+        acc1Modified.data.account().balance = 5000;
+
+        // Add modified acc1 as LIVE entry (shadows the INIT entry)
+        lm.setNextLedgerEntryBatchForBucketTesting(
```

Comment:
I don't think this will actually create a shadow, since we'll just merge into the current snap immediately. I think if we close a few more ledgers in between the INIT and LIVE events (like 4 ledgers) we can guarantee an actual shadow.

Reply:
Ah good point. Updated.